README.md
# @goa/http-errors

[![npm version](https://badge.fury.io/js/%40goa%2Fhttp-errors.svg)](https://www.npmjs.com/package/@goa/http-errors)

`@goa/http-errors` is Creates HTTP Errors For Goa Apps.

```sh
yarn add @goa/http-errors
```

## Table Of Contents

- [Table Of Contents](#table-of-contents)
- [API](#api)
  * [`HttpError`](#type-httperror)
- [`httpErrors(status: number, message: string, props: string): !Error`](#httperrorsstatus-numbermessage-stringprops-string-error)
- [new createError.ErrorType([msg]))](#new-createerrorerrortypemsg)
- [Copyright & License](#copyright--license)

<p align="center"><a href="#table-of-contents">
  <img src="/.documentary/section-breaks/0.svg?sanitize=true">
</a></p>

## API

The package is available by importing its default function and named class:

```js
import httpErrors, { HTTPError } from '@goa/http-errors'
```

__<a name="type-httperror">`HttpError`</a>__: The error constructor that extends Error.

|      Name      |       Type       |                                                                      Description                                                                       |
| -------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| __status__     | <em>number</em>  | The status message.                                                                                                                                    |
| __statusCode__ | <em>string</em>  | The status code.                                                                                                                                       |
| __headers__    | <em>*</em>       | Can be an object of header names to values to be sent to the client, defaulting to `undefined`. When defined, the key names should all be lower-cased. |
| __message__    | <em>string</em>  | The traditional error message, which should be kept short and all single line.                                                                         |
| __expose__     | <em>boolean</em> | Whether to expose the error to the client.<br/>For client errors the default is `true`, for server errors (status &gt;= 500) is `false`.               |

<p align="center"><a href="#table-of-contents">
  <img src="/.documentary/section-breaks/1.svg?sanitize=true">
</a></p>

## <code><ins>httpErrors</ins>(</code><sub><br/>&nbsp;&nbsp;`status: number,`<br/>&nbsp;&nbsp;`message: string,`<br/>&nbsp;&nbsp;`props: string,`<br/></sub><code>): <i>!Error</i></code>
Create a new error object with the given message msg. The error object inherits from createError.HttpError.

 - <kbd><strong>status*</strong></kbd> <em>`number`</em>: The status code as number.
 - <kbd><strong>message*</strong></kbd> <em>`string`</em>: The message. By default, will look up in the status code table.
 - <kbd><strong>props*</strong></kbd> <em>`string`</em>: Additional custom properties to attach to object.

```js
import { aqt } from 'rqt'
import httpErrors from '@goa/http-errors'
import Goa from '@goa/koa'

const goa = new Goa()

goa.use(() => {
  throw httpErrors(401, 'Please login to view this page.')
})
goa.listen(async function() {
  const url = `http://localhost:${this.address().port}`
  const res = await aqt(url)
  console.log(res)
  this.close()
})
```
```js
{ body: 'Please login to view this page.',
  headers: 
   { 'content-type': 'text/plain; charset=utf-8',
     'content-length': '31',
     date: 'Sat, 14 Dec 2019 18:32:50 GMT',
     connection: 'close' },
  statusCode: 401,
  statusMessage: 'Unauthorized' }
```

Another example that extends the given error. _Koa_ will automatically set status **404** for errors with `ENOENT` code.

```js
import { aqt } from 'rqt'
import { readFile } from 'fs'
import createError from '@goa/http-errors'
import Goa from '@goa/koa'
import { join } from 'path'

const goa = new Goa()

goa.use(async (ctx) => {
  await new Promise((r, j) => {
    readFile(join('example', ctx.path), (err) => {
      let httpError
      if (err.code == 'ENOENT') {
        httpError = createError(404, err, { expose: false })
      } else {
        httpError = createError(500, err)
      }
      j(httpError)
    })
  })
})
goa.listen(async function() {
  // 404
  console.log('Request missing file')
  let url = `http://localhost:${this.address().port}/missing.txt`
  let res = await aqt(url)
  console.log(res)
  // 500
  console.log('\nRequest a dir')
  url = `http://localhost:${this.address().port}/dir`
  res = await aqt(url)
  console.log(res)
  this.close()
})
```
```js
Request missing file
{ body: 'Not Found',
  headers: 
   { 'content-type': 'text/plain; charset=utf-8',
     'content-length': '9',
     date: 'Sat, 14 Dec 2019 18:32:50 GMT',
     connection: 'close' },
  statusCode: 404,
  statusMessage: 'Not Found' }

Request a dir
{ body: 'Internal Server Error',
  headers: 
   { 'content-type': 'text/plain; charset=utf-8',
     'content-length': '21',
     date: 'Sat, 14 Dec 2019 18:32:50 GMT',
     connection: 'close' },
  statusCode: 500,
  statusMessage: 'Internal Server Error' }
```

The app will write to _stderr_ on internal error:

```
Error: EISDIR: illegal operation on a directory, read
```

<p align="center"><a href="#table-of-contents">
  <img src="/.documentary/section-breaks/2.svg?sanitize=true">
</a></p>

## new createError.ErrorType([msg]))

A new error could be created from a name or code, like so:

```js
import createError from '@goa/http-errors'

const err = new createError.NotFound()
console.log(err)
```
```js
{ NotFoundError: Not Found
    at Object.<anonymous> (example/constructor.js:3:13)
    at Module._compile (module.js:653:30)
    at Module.p._compile (node_modules/documentary/node_modules/alamode/compile/depack.js:49:18)
    at Module._extensions..js (module.js:664:10)
    at Object.k.(anonymous function).y._extensions.(anonymous function) [as .js] (node_modules/documentary/node_modules/alamode/compile/depack.js:51:7)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
  message: 'Not Found',
  status: 404,
  statusCode: 404,
  expose: true,
  headers: null,
  name: 'NotFoundError' }
```

It's not possible to import specific errors as they are properties on the exported function, and not exports themselves.

**List of all constructors**

|Status Code|Constructor Name             |
|-----------|-----------------------------|
|400        |BadRequest                   |
|401        |Unauthorized                 |
|402        |PaymentRequired              |
|403        |Forbidden                    |
|404        |NotFound                     |
|405        |MethodNotAllowed             |
|406        |NotAcceptable                |
|407        |ProxyAuthenticationRequired  |
|408        |RequestTimeout               |
|409        |Conflict                     |
|410        |Gone                         |
|411        |LengthRequired               |
|412        |PreconditionFailed           |
|413        |PayloadTooLarge              |
|414        |URITooLong                   |
|415        |UnsupportedMediaType         |
|416        |RangeNotSatisfiable          |
|417        |ExpectationFailed            |
|418        |ImATeapot                    |
|421        |MisdirectedRequest           |
|422        |UnprocessableEntity          |
|423        |Locked                       |
|424        |FailedDependency             |
|425        |UnorderedCollection          |
|426        |UpgradeRequired              |
|428        |PreconditionRequired         |
|429        |TooManyRequests              |
|431        |RequestHeaderFieldsTooLarge  |
|451        |UnavailableForLegalReasons   |
|500        |InternalServerError          |
|501        |NotImplemented               |
|502        |BadGateway                   |
|503        |ServiceUnavailable           |
|504        |GatewayTimeout               |
|505        |HTTPVersionNotSupported      |
|506        |VariantAlsoNegotiates        |
|507        |InsufficientStorage          |
|508        |LoopDetected                 |
|509        |BandwidthLimitExceeded       |
|510        |NotExtended                  |
|511        |NetworkAuthenticationRequired|

## Copyright & License

GNU Affero General Public License v3.0

[Original work](https://github.com/jshttp/http-errors) by Jonathan Ong and Douglas Christopher Wilson under MIT license found in [COPYING](COPYING).

<table>
  <tr>
    <td><img src="https://avatars2.githubusercontent.com/u/40834161?s=100&amp;v=4" alt="idiocc"></td>
    <td>© <a href="https://www.idio.cc">Idio™</a> 2019</td>
  </tr>
</table>

<p align="center"><a href="#table-of-contents">
  <img src="/.documentary/section-breaks/-1.svg?sanitize=true">
</a></p>

CHANGELOG.md
## 14 December 2019

### [1.0.0](https://github.com/idiocc/http-errors/compare/v0.0.0-pre...v1.0.0)

- [package] Publish the package.

### 0.0.0-pre

- Create `@goa/http-errors` with _[`My New Package`](https://MNPJS.org)_
- [repository]: `src`, [`test`](https://contexttesting.com), [`documentary`](https://readme.page) & [`types`](https://typedef.page).

compile/http-errors.js
'use strict';var f={[100]:"Continue",[101]:"Switching Protocols",[102]:"Processing",[103]:"Early Hints",[200]:"OK",[201]:"Created",[202]:"Accepted",[203]:"Non-Authoritative Information",[204]:"No Content",[205]:"Reset Content",[206]:"Partial Content",[207]:"Multi-Status",[208]:"Already Reported",[226]:"IM Used",[300]:"Multiple Choices",[301]:"Moved Permanently",[302]:"Found",[303]:"See Other",[304]:"Not Modified",[305]:"Use Proxy",[306]:"(Unused)",[307]:"Temporary Redirect",[308]:"Permanent Redirect",[400]:"Bad Request",
[401]:"Unauthorized",[402]:"Payment Required",[403]:"Forbidden",[404]:"Not Found",[405]:"Method Not Allowed",[406]:"Not Acceptable",[407]:"Proxy Authentication Required",[408]:"Request Timeout",[409]:"Conflict",[410]:"Gone",[411]:"Length Required",[412]:"Precondition Failed",[413]:"Payload Too Large",[414]:"URI Too Long",[415]:"Unsupported Media Type",[416]:"Range Not Satisfiable",[417]:"Expectation Failed",[418]:"I'm a teapot",[421]:"Misdirected Request",[422]:"Unprocessable Entity",[423]:"Locked",
[424]:"Failed Dependency",[425]:"Unordered Collection",[426]:"Upgrade Required",[428]:"Precondition Required",[429]:"Too Many Requests",[431]:"Request Header Fields Too Large",[451]:"Unavailable For Legal Reasons",[500]:"Internal Server Error",[501]:"Not Implemented",[502]:"Bad Gateway",[503]:"Service Unavailable",[504]:"Gateway Timeout",[505]:"HTTP Version Not Supported",[506]:"Variant Also Negotiates",[507]:"Insufficient Storage",[508]:"Loop Detected",[509]:"Bandwidth Limit Exceeded",[510]:"Not Extended",
[511]:"Network Authentication Required"};/*
 statuses
 Copyright(c) 2014 Jonathan Ong
 Copyright(c) 2016 Douglas Christopher Wilson
 MIT Licensed
*/
const l=k();function k(){var a=m;const b=[];Object.keys(f).forEach(d=>{const c=f[d];d=Number(d);a[d]=c;a[c]=d;a[c.toLowerCase()]=d;b.push(d)});return b}function m(a){if("number"==typeof a){if(!m[a])throw Error("invalid status code: "+a);return a}if("string"!=typeof a)throw new TypeError("code must be a number or string");var b=parseInt(a,10);if(!isNaN(b)){if(!m[b])throw Error("invalid status code: "+b);return b}b=m[a.toLowerCase()];if(!b)throw Error('invalid status message: "'+a+'"');return b};/*
 http-errors
 Copyright(c) 2014 Jonathan Ong
 Copyright(c) 2016 Douglas Christopher Wilson
 MIT Licensed
 toidentifier
 Copyright(c) 2016 Douglas Christopher Wilson
 MIT Licensed
*/
function n(...a){let b,d,c=500,g={};for(let e=0;e<a.length;e++){let h=a[e];if(h instanceof Error)b=h,c=b.status||b.statusCode||c;else switch(typeof h){case "string":d=h;break;case "number":c=h;0!==e&&process.emitWarning("non-first-argument status code; replace with createError("+h+", ...)","DeprecationWarning");break;case "object":g=h}}"number"==typeof c&&(400>c||600<=c)&&process.emitWarning("non-error status code; use only 4xx or 5xx status codes","DeprecationWarning");if("number"!=typeof c||!m[c]&&
(400>c||600<=c))c=500;a=n[c]||n[Number(String(c).charAt(0)+"00")];b||(b=a?new a(d):Error(d||m[c]),Error.captureStackTrace(b,n));a&&b instanceof a&&b.status===c||(b.expose=500>c,b.status=b.statusCode=c);for(let e in g)"status"!=e&&"statusCode"!=e&&(b[e]=g[e]);return b}class p extends Error{constructor(a){super();this.message=a;this.statusCode=this.status=null;this.expose=!1;this.headers=null}set code(a){this.statusCode=this.status=a;this.message||(this.message=m[a])}}
l.forEach(a=>{let b;const d=q(m[a]),c=d.match(/Error$/)?d:d+"Error";switch(Number(String(a).charAt(0)+"00")){case 400:b=class extends p{constructor(g){super(g);this.code=a;this.name=c;this.expose=!0}};break;case 500:b=class extends p{constructor(g){super(g);this.code=a;this.name=c;this.expose=!1}}}b&&(n[a]=b,n[d]=b)},{});function q(a){return a.split(" ").map(function(b){return b.charAt(0).toUpperCase()+b.slice(1)}).join("").replace(/[^ _0-9a-z]/gi,"")};module.exports={_HttpError:p,_httpErrors:n};

//# sourceMappingURL=http-errors.js.map